mithril-common/src/resource_pool.rs

```diff
@@ -0,0 +1,320 @@
+    discriminant: Mutex<u64>,
+
+    /// Resources in the pool
+    resources: Mutex<VecDeque<T>>,
+
+    /// Condition variable to notify when a resource is available
+    not_empty: Condvar,
+}
+
+impl<T: Send + Sync> ResourcePool<T> {
+    /// Create a new resource pool
+    pub fn new(pool_size: usize, resources: Vec<T>) -> Self {
+        Self {
+            size: pool_size,
+            discriminant: Mutex::new(0),
+            resources: Mutex::new(resources.into()),
+            not_empty: Condvar::new(),
+        }
+    }
+
+    /// Acquire a resource from the pool with a timeout
+    pub fn acquire_resource(&self, timeout: Duration) -> StdResult<ResourcePoolItem<'_, T>> {
+        let mut resources = self
+            .resources
+            .lock()
+            .map_err(|_| ResourcePoolError::PoisonedLock())
+            .with_context(|| "Resource pool 'acquire_resource' failed locking Mutex")?;
+        while resources.is_empty() {
+            let (resources_locked, wait_result) = self
+                .not_empty
+                .wait_timeout(resources, timeout)
+                .map_err(|_| ResourcePoolError::PoisonedLock())
+                .with_context(|| "Resource pool 'acquire_resource' failed waiting for resource")?;
+            if wait_result.timed_out() {
+                return Err(ResourcePoolError::AcquireTimeout())
+                    .with_context(|| "Resource pool 'acquire_resource' has timed out");
+            }
+            resources = resources_locked;
+        }
+
+        Ok(ResourcePoolItem::new(self, resources.pop_front().unwrap()))
+    }
+
+    /// Give back a resource to the pool
+    /// A resource is given back to the pool only if the discriminant matches
+    /// and if the pool is not already full
+    pub fn give_back_resource(&self, resource: T, discriminant: u64) -> StdResult<()> {
+        if self.count()? == self.size {
+            // Pool is full
+            return Ok(());
+        }
+        let mut resources = self
+            .resources
+            .lock()
+            .map_err(|_| ResourcePoolError::PoisonedLock())
+            .with_context(|| "Resource pool 'give_back_resource' failed locking Mutex")?;
+        if self.discriminant()? != discriminant {
+            // Stale resource
+            return Ok(());
+        }
+        resources.push_back(resource);
+        self.not_empty.notify_one();
+
+        Ok(())
+    }
+
+    /// Clear the pool
+    pub fn clear(&self) {
+        let mut resources = self.resources.lock().unwrap();
+        resources.clear();
+    }
+
+    /// Get the discriminant of the resource pool item
+    pub fn discriminant(&self) -> StdResult<u64> {
+        Ok(*self
+            .discriminant
+            .lock()
+            .map_err(|_| ResourcePoolError::PoisonedLock())
+            .with_context(|| "Resource pool 'discriminant' failed locking Mutex")?)
+    }
+
+    /// Set the discriminant of the resource pool item
```

Comment:
We may create a function to lock the Mutex and hide the error mapping that is done several times.

```
pub fn set_discriminant(&self, discriminant: u64) -> StdResult<()> {
    let mut discriminant_guard = self.lock(&self.discriminant, "set_discriminant")?;
    *discriminant_guard = discriminant;

    Ok(())
}

fn lock<'a, V>(
    &self,
    item: &'a Mutex<V>,
    current_function: &str,
) -> Result<std::sync::MutexGuard<'a, V>, anyhow::Error> {
    Ok(item
        .lock()
        .map_err(|_| ResourcePoolError::PoisonedLock())
        .with_context(|| {
            format!("Resource pool '{}' failed locking Mutex", current_function)
        })?)
}
```